x/cork/keeper/keeper.go

```diff
@@ -111,6 +113,85 @@ func (k Keeper) IterateCorks(ctx sdk.Context, handler func(val sdk.ValAddress, c
 	}
 }
 
+/////////////////////
+// Scheduled Corks //
+/////////////////////
+
+func (k Keeper) SetScheduledCork(ctx sdk.Context, val sdk.ValAddress, blockHeight uint64, cork types.Cork) {
+	bz := k.cdc.MustMarshal(&cork)
+	ctx.KVStore(k.storeKey).Set(types.GetScheduledCorkKey(val, blockHeight, common.HexToAddress(cork.TargetContractAddress)), bz)
+}
+
+// GetScheduledCork gets the scheduled cork for a given validator
+// CONTRACT: must provide the validator address here not the delegate address
+func (k Keeper) GetScheduledCork(ctx sdk.Context, val sdk.ValAddress, blockHeight uint64, contract common.Address) (types.Cork, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetScheduledCorkKey(val, blockHeight, contract))
+	if len(bz) == 0 {
+		return types.Cork{}, false
+	}
+
+	var cork types.Cork
+	k.cdc.MustUnmarshal(bz, &cork)
+	return cork, true
+}
+
+// DeleteScheduledCork deletes the scheduled cork for a given validator
+// CONTRACT: must provide the validator address here not the delegate address
+func (k Keeper) DeleteScheduledCork(ctx sdk.Context, val sdk.ValAddress, blockHeight uint64, contract common.Address) {
+	ctx.KVStore(k.storeKey).Delete(types.GetScheduledCorkKey(val, blockHeight, contract))
+}
+
+// HasScheduledCork gets the existence of any scheduled commit for a given validator
+// CONTRACT: must provide the validator address here not the delegate address
+func (k Keeper) HasScheduledCork(ctx sdk.Context, val sdk.ValAddress) bool {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.GetScheduledCorkValidatorKeyPrefix(val))
+	iter := store.Iterator(nil, nil)
+	defer iter.Close()
+
+	return iter.Valid()
+}
+
+// IterateScheduledCorks iterates over all scheduled corks in the store
+func (k Keeper) IterateScheduledCorks(ctx sdk.Context, cb func(val sdk.ValAddress, blockHeight uint64, cel common.Address, cork types.Cork) (stop bool)) {
+	store := ctx.KVStore(k.storeKey)
+	iter := sdk.KVStorePrefixIterator(store, []byte{types.ScheduledCorkForAddressKey})
+	defer iter.Close()
+
+	for ; iter.Valid(); iter.Next() {
+		var cork types.Cork
+		keyPair := bytes.NewBuffer(bytes.TrimPrefix(iter.Key(), []byte{types.ScheduledCorkForAddressKey}))
+		val := sdk.ValAddress(keyPair.Next(20))
+		blockHeight := binary.BigEndian.Uint64(keyPair.Next(8))
+		cel := common.BytesToAddress(keyPair.Bytes())
+
+		k.cdc.MustUnmarshal(iter.Value(), &cork)
+		if cb(val, blockHeight, cel, cork) {
+			break
+		}
+	}
+}
+
+func (k Keeper) ScheduledBlockHeights(ctx sdk.Context) []uint64 {
+	blockHeightsMap := make(map[uint64]bool)
+	k.IterateScheduledCorks(ctx, func(_ sdk.ValAddress, blockHeight uint64, _ common.Address, _ types.Cork) (stop bool) {
```

Comment:
It seems expensive to iterate and sort over every scheduled cork to determine if any exist for a given block height. Is there a more efficient way to store a mapping of block heights that indicate there are scheduled corks processable for that height? You could do something like normalize block heights when storing/processing them to vote period boundaries so you can check a constant value for a given current block height.

Reply:
A second question here...is anything actually calling this function?

Reply:
No one is currently using this function, it was potentially for a query function, which I am not sure is useful for anyone